core/src/main/java/org/elasticsearch/cluster/metadata/IndexGraveyard.java

```diff
@@ -0,0 +1,449 @@
+
+    // holds the index tombstones
+    private final Deque<Tombstone> tombstones;
+
+    public IndexGraveyard() {
+        tombstones = new ArrayDeque<>();
+    }
+
+    // copy constructor
+    public IndexGraveyard(final IndexGraveyard that) {
+        this.tombstones = new ArrayDeque<>(that.tombstones);
+    }
+
+    public IndexGraveyard(final StreamInput in) throws IOException {
+        final int queueSize = in.readVInt();
+        tombstones = new ArrayDeque<>(queueSize);
+        for (int i = 0; i < queueSize; i++) {
+            tombstones.add(new Tombstone(in));
+        }
+    }
+
+    // only used by fromXContent
+    private IndexGraveyard(final List<Tombstone> list) {
+        this.tombstones = new ArrayDeque<>(list);
+    }
+
+    // only used by IndexGraveyardDiff
+    private IndexGraveyard(final Deque<Tombstone> queue) {
+        this.tombstones = new ArrayDeque<>(queue);
+    }
+
+    /**
+     * Do they have the same index tombstones (deletions)?
+     */
+    public boolean isEquals(IndexGraveyard that) {
+        // array deque doesn't implement equals
+        return Objects.equals(new ArrayList<>(tombstones), new ArrayList<>(that.tombstones));
+    }
+
+    /**
+     * Add a deleted index to the list of tombstones in the cluster state.  Returns the
+     * purged tombstone if the size of the list grew too big, else return none.
+     */
+    public Optional<Tombstone> addTombstone(final Index index) {
+        return addTombstone(index, System.currentTimeMillis(), System.nanoTime());
+    }
+
+    /**
+     * Add a deleted index to the list of tombstones in the cluster state, along with
+     * a specified deletion date and nanosecond date.  Returns the purged tombstone
+     * if the size of the list grew too big, else return none.
+     */
+    Optional<Tombstone> addTombstone(final Index index, final long deletionDate, final long deletionDateNS) {
+        final Optional<Tombstone> purged;
+        if (tombstones.size() == MAX_TOMBSTONES) {
+            // purge the oldest entry
+            purged = Optional.of(tombstones.removeFirst());
+        } else {
+            purged = Optional.empty();
+        }
+        tombstones.addLast(new Tombstone(index, deletionDate, deletionDateNS));
+        return purged;
+    }
+
+    /**
+     * Get a copy of the current index tombstones.
+     */
+    List<Tombstone> getTombstones() {
+        return new ArrayList<>(tombstones);
+    }
+
+    /**
+     * Purge tombstone entries by date, if they have passed a certain expiration window.
+     * Returns the number of entries that were purged.
+     */
+    public int purge() {
+        int count = 0;
+        final long now = System.currentTimeMillis();
+        while (tombstones.isEmpty() == false) {
+            // start from the beginning, it has the oldest entries
+            final Tombstone tombstone = tombstones.getFirst();
+            if (now - tombstone.deleteDate > EXPIRATION_WINDOW) {
```

Comment:
I like this idea, because it preserves the absolute nano time **difference** across master nodes as master changes hands, with the minor added complexity of updating the time difference on each cluster state update.  I'm in favor of this.